src/Raven.Server/Web/System/AdminDatabasesHandler.cs

```diff
@@ -762,72 +772,95 @@ public async Task Delete()
+                await using ( var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
                 {
                     context.Write(writer, new DynamicJsonValue
                     {
                         // we only send the successful index here, we might fail to delete the index
                         // because a node is down, and we don't want to cause the client to wait on an
                         // index that doesn't exists in the Raft log
-                        [nameof(DeleteDatabaseResult.RaftCommandIndex)] = actualDeletionIndex,
+                        [nameof(DeleteDatabaseResult.RaftCommandIndex)] = lastRemoveIndex,
                         [nameof(DeleteDatabaseResult.PendingDeletes)] = new DynamicJsonArray(pendingDeletes)
                     });
                 }
             }
         }
 
+        private async Task<long> WaitForRemovingNodesFromDatabases(
+            Dictionary<string, DeleteDatabaseDetails> databasesToDelete,
+            long lastDeletionIndex,
+            TimeSpan timeout)
+        {
+            if (databasesToDelete.Count == 0)
+                return -1;
+
+            await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, lastDeletionIndex);
+
+            long lastRemoveIndex = 0;
+
+            var reqExecutors = new Dictionary<string, ClusterRequestExecutor>();
+            try
+            {
+                foreach (var (dbName, details) in databasesToDelete)
+                {
+                    var deleteCmdIndex = details.RaftIndex;
+
+                    foreach (var removedNodeTag in details.NodesToRemove) // nodeTags
+                    {
+                        var removeNodeFromDbCmdRaftIndex = 0L;
+                        var sw = Stopwatch.StartNew();
+                        while (TryGetLastRemoveNodeFromDatabaseCommand(ServerStore, dbName, removedNodeTag, deleteCmdIndex, out removeNodeFromDbCmdRaftIndex) == false)
+                        {
+                            if (sw.Elapsed > timeout)
+                            {
+                                throw new InvalidOperationException($"RemoveNodeFromDatabase command of node {removedNodeTag} doesn't exist in the log of node {ServerStore.NodeTag}.");
+                            }
+
+                            await Task.Delay(TimeSpan.FromMilliseconds(300));
+                        }
+
+                        lastRemoveIndex = long.Max(lastRemoveIndex, removeNodeFromDbCmdRaftIndex);
+                    }
+                }
+
+                // RemoveNodeFromDatabase command is being created in the database (on the node DatabaseLandlord) after the db has been deleted from it.
+                using (var cts = new CancellationTokenSource(timeout))
+                    await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, lastRemoveIndex, cts.Token);
```

Comment:
this will throw operation cancelled exception, right? if so it is misleading and should throw timeout exception.

Reply:
No, because the timeout exception will cause the client request-executor to perform a failover, which I don't want

Reply:
you still may throw timeout exception here, so we must ensure that we handle the failover properly as well